### ISSUE TITLE
Progress bar?

### ISSUE THREAD

**Issue by HenKlei** [pymor/pymor#953]
For some tasks that take a long time to complete or that consist of many steps, a progress bar might be helpful. Always using the logging mechanism for that purpose could lead to massive logging output without much helpful information.

**Comment by renefritze** [comment]
I know I added a progress bar for use in notebooks. But it seems unused currently. I'm also pretty sure there was a terminal based one at some point, but I'd have go hunting git history. 

https://github.com/pymor/pymor/blob/e55fe2c85710665f19305b8460fa6b3dde4e8ee0/src/pymor/discretizers/builtin/gui/jupyter/__init__.py#L29
Should be used like `range` to wrap a sequence to iterate over, definitely needs a docstring.

**Comment by sdrave** [comment]
I think this needs a broader discussion. Maybe we should look for places in pyMOR where this would be useful. Regarding implementations, I myself have enjoyed using [tqdm](https://github.com/tqdm/tqdm), which supposedly also works with jupyter.

**Comment by HenKlei** [comment]
Using `tqdm` would require installing another package. Maybe we can write a less advanced and simpler progress bar instead of relying on additional external code.

**Comment by sdrave** [comment]
I would assume that it is quite non-trivial to write a progress bar that reliably works in all terminals. If it's more like 20 lines of code, I see no reason to reinvent the wheel here. Also, tqdm has 15k stars on github. It's not going away any time soon.

**Comment by HenKlei** [comment]
That's true.

Maybe we can make `tqdm` optional and not mandatory?

**Comment by sdrave** [comment]
> Maybe we can make tqdm optional and not mandatory?

How would that work. You mean that the entire progress bar is optional and does not appearch when tqdm is not installed, or do you want an alternative implementation for that case?

**Comment by HenKlei** [comment]
Right, if `tqdm` is not installed, no progress bar shows up. This way, we would not force installing `tqdm`, but provide a progress bar in the case it is installed.

**Comment by renefritze** [comment]
I think we should be careful with using progressbars in the library proper. I often have to look at combined stderr/out stream "logs" from a queue system for example. These would look horrible with progress bars in them. Same for CI logs. So if they were to be used, I must be able to turn them off at least globally (forcing me not to have tqdm installed is not an option). 

Usage in in demos/tutorials is ofc fine and I think a great idea.

**Comment by sdrave** [comment]
@renefritze do you know how tqdm behaves in this case? Maybe it is smart enough to see that it is not connected to a proper terminal?

> Usage in in demos/tutorials is ofc fine and I think a great idea.

I think it is more interesting to discuss whether or not to use this inside of pyMOR's algorithms (e.g. `gram_schmidt`).

> Right, if tqdm is not installed, no progress bar shows up. This way, we would not force installing tqdm, but provide a progress bar in the case it is installed.

That would be a possibility, but also some  extra work since we would need to duplicate all parts of tqdm's API we want to use. If it is really needed to disable the functionality (e.g. for @renefritze's use case) that would be reasonable. I would not do it just to make one dependency optional.

**Comment by sdrave** [comment]
Regarding redirection see [Using tqdm when redirecting output?](https://github.com/tqdm/tqdm/issues/506). Sounds like @renefritze will not think of this as a satisfying solution?

**Comment by renefritze** [comment]
> Maybe it is smart enough to see that it is not connected to a proper terminal?

At least with slurm it would actually be connected to a proper terminal. Ofc it could still be smart enough to detect the runtime env through some other means. I haven't actually used tqdm at all. 

> Sounds like @renefritze will not think of this as a satisfying solution?

Correct. Most of the time I do not use redirection inside the queue script since the manager does that for me, and better. With per Rank logfiles and such.




> I would not do it just to make one dependency optional.

Agreed.  It's a pure python package with no external deps itself too. So I think the barrier for extra work to keep tqdm entirely optional is super low.

**Comment by HenKlei** [comment]
> That would be a possibility, but also some extra work since we would need to duplicate all parts of tqdm's API we want to use.

Why do we need to duplicate code of `tqdm` in that case? Couldn't we just check whether `tqdm` is installed as we do with, for instance, FEniCS and use a progress bar if `tqdm` is available?

**Comment by renefritze** [comment]
Our to-be tdqm shim would ne to be able to accept all calls tdqm can. Otherwise we'd need to implement that check (and fork code paths accordingly) everywhere we want to use tqdm instead of just once.

**Comment by HenKlei** [comment]
Alright, I see.

**Comment by HenKlei** [comment]
Typer seems to also support progressbars: https://typer.tiangolo.com/tutorial/progressbar/

**Comment by sdrave** [comment]
> Typer seems to also support progressbars: https://typer.tiangolo.com/tutorial/progressbar/

However, it does not seem to be able to estimate time till completion. I think this is a very useful feature tqdm has ..

**Comment by renefritze** [comment]
I'll throw the [rich implementation](https://rich.readthedocs.io/en/stable/progress.html) into the ring.
pip's [using it](https://github.com/pradyunsg/pip/blob/main/src/pip/_internal/cli/progress_bars.py#L273) since version 22
It looks good in [logfiles](https://zivgitlab.uni-muenster.de/pymor/docker/-/jobs/814682#L436). 
I like rich :wink:

**Comment by ftalbrecht** [comment]
> I'll throw the [rich implementation](https://rich.readthedocs.io/en/stable/progress.html) into the ring. pip's [using it](https://github.com/pradyunsg/pip/blob/main/src/pip/_internal/cli/progress_bars.py#L273) since version 22 It looks good in [logfiles](https://zivgitlab.uni-muenster.de/pymor/docker/-/jobs/814682#L436). I like rich wink

Looks, good! In particular the latter  is appealing!

**Comment by HenKlei** [comment]
A first version using rich is implemented [here](https://github.com/pymor/pymor/tree/progress-bar-neural-network-training) for the neural network training. You can check it out by running one of the neural network demos.

**Comment by sdrave** [comment]
Done in #2552.
